sql/expression/function/json_set.go

```diff
@@ -188,31 +186,129 @@ func (j *JSONSet) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
 // processPath checks the given json path for the correct mysql syntax, checks nested paths for their existence
 // in the provided json doc, determines whether JSON_SET will do nothing with no error for this path, and processes
 // the given json path to use the appropriate sjson syntax.
-func processPath(ctx *sql.Context, doc types.JSONDocument, path string, pass bool) (string, bool, error) {
-	// make sure path starts with '$'
-	if path[0] != '$' {
-		return "", false, fmt.Errorf("Invalid JSON path expression")
+func processPath(doc, path string, returnVal bool) (string, bool, bool, error) {
+	returnVal, err := checkPath(path)
+	if err != nil {
+		return "", false, false, err
 	}
+	if returnVal {
+		return "", false, true, nil
+	}
+	path = path[1:]
 
-	// no wildcards allowed in path
-	if strings.Contains(path, "*") || strings.Contains(path, "**") {
-		return "", false, fmt.Errorf("Path expressions may not contain the * and ** tokens")
+	// tokenize each field of the path
+	var parsed parsedPath
+	parsed.parts = strings.Split(path, ".")
+	if parsed.parts[0] == "" {
+		parsed.parts = parsed.parts[1:]
 	}
 
-	// if the nested path does not already exist, we do nothing
-	if len(strings.Split(path, ".")) > 2 {
-		extractedVal, _ := doc.Extract(ctx, path)
-		if extractedVal == nil {
-			pass = true
+	// process each field of the path
+	for i, part := range parsed.parts {
+		formattedPart := part
+
+		// handle any indexing in this field
+		if strings.Contains(part, "[") {
+			// tokenize indexes
+			formattedPart = strings.ReplaceAll(formattedPart, "]", "")
+			indexedParts := strings.Split(formattedPart, "[")
+			if indexedParts[0] == "" {
+				indexedParts = indexedParts[1:]
+			}
+			formattedPart = ""
+
+			// process each token
+			for j, idxPart := range indexedParts {
+				if idx, err := strconv.Atoi(idxPart); err == nil { // if token is an int, it's an index
+					if j == 0 {
+						if i == 0 {
+							return "", false, false, fmt.Errorf("ordinal indexing currently unsupported")
+						} else {
+							return "", false, false, fmt.Errorf("Invalid JSON path expression")
+						}
+					}
+
+					parentVal := gjson.Get(doc, parsed.formattedPath+formattedPart)
+
+					if !parentVal.Exists() { // if parent doesn't exist in json already, do nothing
+						return path, true, false, nil
+					}
```

Comment:
Moving comments above the if statements will probably make this a little easier to read. 
```suggestion
					// if parent doesn't exist in json already, do nothing
					if !parentVal.Exists() { 
						return path, true, false, nil
					}

```